src/bindtextdomain.c

```diff
@@ -0,0 +1,106 @@
+/* gcc -Wall -O2 -o bindtextdomain.so -fPIC -shared bindtextdomain.c -ldl */
+
+#define _GNU_SOURCE
+
+#include <errno.h>
+#include <dirent.h>
+#include <dlfcn.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+static void die (const char *msg)
+{
+        fprintf (stderr, "can't preload: %s\n", msg);
+        exit (EXIT_FAILURE);
+}
+
+typedef char * (*BindTextDomainFunc) (const char *a,
+                                      const char *b);
+
+static BindTextDomainFunc r_bindtextdomain = NULL;
+
+char *bindtextdomain(const char *domainname, const char *dirname)
+{
+        char *snap = NULL;
+        char *snap_path = NULL;
+        char *snap_locale_path = NULL;
+        DIR *dir = NULL;
+        struct dirent *ent = NULL;
+        char *ret = NULL;
+
+        if (r_bindtextdomain == 0 && !(r_bindtextdomain =
+                                       (BindTextDomainFunc) dlsym (RTLD_NEXT, "bindtextdomain")))
+                die ("can't find symbol \"bindtextdomain\"");
+
+        if (dirname == NULL || dirname[0] != '/')
+                goto orig;
+
+        snap = getenv ("SNAP");
+
+        if (snap == NULL || strcmp (snap, "") == 0)
+                goto orig;
+
+	char * paths[] = {
+		"gnome-platform/usr/share/locale",
+		"usr/share/locale",
+		"usr/share/locale-langpack",
+		NULL
+	};
+	for(int i = 0; paths[i] != NULL; i++)
+	{
+		if (asprintf (&snap_path, "%s/%s", snap, paths[i]) < 0)
+			continue;
+
+		if (access (snap_path, F_OK) < 0)
+			continue;
```

Comment:
Shouldn't snap_path (and snap_locale_path) be freed at every iteration to avoid leaking memory potentially allocated by asprintf?

Reply:
Good catch, I think I've handled that now.